File: src/aiusage/diagnostics.py
import datetime as dt
import locale
import platform

from . import __version__
from .providers import REGISTRY
from .timezones import convert, offset_label


def collect(cfg, github_ok=None):
    rows = [
        ("AIUsage", True, f"v{__version__}"),
        ("Python", tuple(map(int, platform.python_version_tuple()[:2])) >= (3, 10), platform.python_version()),
        ("Terminal", bool((locale.getpreferredencoding(False) or "").lower().replace("-", "").startswith("utf8")), locale.getpreferredencoding(False)),
        ("Config", True, "readable"),
    ]
    for key in ("codex", "grok"):
        adapter = REGISTRY[key]
        installed = adapter.installed()
        rows.append((adapter.name, installed, "installed" if installed else "not installed"))
        state = adapter.read() if installed else None
        readable = bool(state and state.windows)
        rows.append((f"{adapter.name} usage", readable, "readable" if readable else "unavailable"))
    instant = dt.datetime.now(dt.timezone.utc)
    now = convert(instant, "system")
    display = convert(instant, cfg.timezone)
    rows.extend([
        ("System timezone", True, offset_label(now)),
        ("Display timezone", True, offset_label(display)),
        ("GitHub", github_ok is True, "available" if github_ok is True else "unknown" if github_ok is None else "unavailable"),
    ])
    return rows


def sanitized_text(rows):
    return "\n".join(f"{name}: {'OK' if ok else 'WARN'} ({detail})" for name, ok, detail in rows)

File: rust/src/diagnostics.rs
use crate::{config::Config, providers, timezones};
pub type Row = (String, bool, String);
pub fn collect(cfg: &Config, github_ok: Option<bool>) -> Vec<Row> {
    let encoding = std::env::var("LC_ALL")
        .or_else(|_| std::env::var("LC_CTYPE"))
        .or_else(|_| std::env::var("LANG"))
        .unwrap_or_default();
    let utf8 = encoding.to_lowercase().replace('-', "").contains("utf8");
    let mut rows = vec![
        ("AIUsage".into(), true, format!("v{}", crate::VERSION)),
        ("Rust".into(), true, "native binary".into()),
        (
            "Terminal".into(),
            utf8,
            if utf8 { "UTF-8" } else { "unknown" }.into(),
        ),
        ("Config".into(), true, "readable".into()),
    ];
    let discovery = providers::discover_all();
    for &(key, name) in crate::PROVIDERS {
        let state = &discovery[key];
        let disabled = cfg.disabled_providers.iter().any(|k| k == key);
        rows.push((
            name.into(),
            state.usable && !disabled,
            if disabled {
                "disabled_by_user".into()
            } else {
                state.reason.clone()
            },
        ));
    }
    for (key, name) in [("codex", "Codex usage"), ("grok", "Grok usage")] {
        let readable = discovery[key].usable && !providers::read(key).windows.is_empty();
        rows.push((
            name.into(),
            readable,
            if readable { "readable" } else { "unavailable" }.into(),
        ));
    }
    let now = crate::cli::now();
    for (name, setting) in [
        ("System timezone", "system"),
        ("Display timezone", cfg.timezone.as_str()),
    ] {
        let value = timezones::from_epoch(now, setting).ok();
        rows.push((
            name.into(),
            value.is_some(),
            value
                .map(|v| timezones::label_for(&v))
                .unwrap_or("unknown".into()),
        ));
    }
    rows.push((
        "GitHub".into(),
        github_ok == Some(true),
        match github_ok {
            Some(true) => "available",
            Some(false) => "unavailable",
            None => "unknown",
        }
        .into(),
    ));
    rows
}
